libpkg/pkg_add.c

```diff
@@ -56,14 +59,137 @@
 #define NOCHANGESFLAGS	(UF_IMMUTABLE | UF_APPEND | SF_IMMUTABLE | SF_APPEND)
 #endif
 
+static merge_status
+merge_with_external_tool(const char *merge_tool, struct pkg_config_file *lcf,
+    size_t lcf_len, struct pkg_config_file *rcf, char *localconf, char **mergedconf)
+{
+	pid_t wait_res;
+	int status;
+	FILE *inout[2];
+
+	char *tmpdir = getenv("TMPDIR");
+	if (tmpdir == NULL)
+		tmpdir = "/tmp";
+
+	int output_fd;
+	char output_path[MAXPATHLEN];
+	off_t output_sz;
+
+	strlcpy(output_path, tmpdir, sizeof(output_path));
+	strlcat(output_path, "/OUTPUT.XXXXXXXXXX", sizeof(output_path));
+	output_fd = mkstemp(output_path);
+	if (output_fd == -1) {
+		pkg_emit_error("Can't create %s", output_path);
+		return MERGE_FAILED;
+	}
+	close(output_fd);
+
+	struct {
+		int fd;
+		const char *template;
+		char path[MAXPATHLEN];
+		const char *content;
+		size_t content_len;
+	} tmp_files[] = {
+		{-1, "/BASE.XXXXXXXXXX", {0}, lcf->content, lcf_len},
+		{-1, "/LOCAL.XXXXXXXXXX", {0}, localconf, strlen(localconf)},
+		{-1, "/REMOTE.XXXXXXXXXX", {0}, rcf->content, strlen(rcf->content)}
+	};
+	bool tmp_files_ok = true;
+	for (int i = 0; i < NELEM(tmp_files); i++) {
+		strlcpy(tmp_files[i].path, tmpdir, sizeof(tmp_files[i].path));
+		strlcat(tmp_files[i].path, tmp_files[i].template, sizeof(tmp_files[i].path));
+
+		tmp_files[i].fd = mkstemp(tmp_files[i].path);
+		if (tmp_files[i].fd == -1) {
+			pkg_emit_error("Can't create %s", tmp_files[i].path);
+			tmp_files_ok = false;
+			break;
+		}
+		if (write(tmp_files[i].fd, tmp_files[i].content, tmp_files[i].content_len) == -1) {
+			pkg_emit_error("Failed to write %s", tmp_files[i].path);
+			tmp_files_ok = false;
+			break;
+		}
+		close(tmp_files[i].fd);
+		tmp_files[i].fd = -1;
+	}
+	if (!tmp_files_ok) {
+		for (int i = 0; i < NELEM(tmp_files); i++) {
+			if (tmp_files[i].fd != -1)
+				close(tmp_files[i].fd);
+			if (strlen(tmp_files[i].path))
+				unlink(tmp_files[i].path);
+		}
+		return MERGE_FAILED;
+	}
+
+	char command[MAXPATHLEN];
+	for (int i = 0; *merge_tool; i++, merge_tool++) {
```

Comment:
style only I like expclicit checks like *merge_tool != '\0' it ease readability for new comers